addons/account/models/account_invoice.py

```diff
@@ -384,13 +399,34 @@ def get_view_id(xid, name):
                     view_id = get_view_id('invoice_form', 'account.invoice.form').id
         return super(AccountInvoice, self).fields_view_get(view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu)
 
+    @api.model
+    def invoice_report_post(self, report):
+        if len(self) > 1:
+            return report
+        if not self.sent and self.state in ('open', 'paid'):
+            attachment_name = 'INV%s.pdf' % self.number.replace('/', '')
+            attachment_id = self.env['ir.attachment'].create({
+                'name': attachment_name,
+                'datas': base64.encodestring(report),
+                'datas_fname': attachment_name,
+                'res_model': 'account.invoice',
+                'res_id': self.id,
+            })
+            self.sent = True
+            # Print on the chatter
+            self.message_post(body=_('Invoice report generated'), attachment_ids=[attachment_id.id], subtype='account.mt_invoice_validated')
+            # Special case when the invoice already has payments
+            if self.payment_ids:
+                ctx = dict(self._context, post_process_report=False)
+                return self.env['report'].with_context(ctx).get_pdf([self.id], 'account.report_invoice')
+        return report
+
     @api.multi
     def invoice_print(self):
         """ Print the invoice and mark it as sent, so that we can see more
             easily the next step of the workflow
         """
         self.ensure_one()
-        self.sent = True
```

Comment:
Because if sent is set at True here, I don't know if the document is generated for the first time or not and then, if I need to add the payments on the pdf.